### PR TITLE
LTD-563: HMRC mock service to simulate reply emails

### PR DESCRIPTION
**Pull request by saruniitr**
## Change description

Adds a mock HMRC service that simulates replies to various licence mails as if they are coming from a real HMRC mailbox.
At the moment only licenceData reply is supported and also cannot simulate all possible error conditions.

WIP.

### CODE REVIEW

File: mock_hmrc/data_processors.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    # ensure there is run number
+    if data["licence_data"]["source_run_number"] is None:
+        logging.error("Invalid email received")
+        update_retrieved_email_status(dto, enums.RetrievedEmailStatusEnum.INVALID)
+        return None
+
+    source_run_number = data["licence_data"]["source_run_number"]
+    try:
+        hmrc_mail = models.HmrcMail.objects.get(extract_type=extract_type, source_run_number=source_run_number)
```
Comment (r4vi) [inline mock_hmrc/data_processors.py:40]: use `get_or_create` instead with `defaults=...`

File: mock_hmrc/data_processors.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        data.append(f"{line_num}\\accepted\\{id}")
+
+    data.append(f"{line_num}\\fileTrailer\\{len(accepted_ids)}\\0\\0")
+    file_data = "\n".join(data)
+
+    return file_data
+
+
+def build_reply_mail_message_dto(mail) -> EmailMessageDto:
+    sender = None
```
[the remark this answers is not in the document]
Comment (saruniitr) [reply mock_hmrc/data_processors.py:84]: This is taken from lite-hmrc, duplicated code, twice not required and will remove this one.

File: mock_hmrc/tasks.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+
+@background(queue="handle_hmrc_replies_queue", schedule=0)
+def handle_replies():
+    logging.info("Polling mock HMRC inbox for updates")
+
+    try:
+        parse_and_reply_emails()
+    except Exception as exc:  # noqa
+        logging.error(f"An unexpected error occurred when polling inbox for updates -> {type(exc).__name__}: {exc}")
```
[the remark this answers is not in the document]
Comment (saruniitr) [reply mock_hmrc/tasks.py:15]: Not sure if the polling continues if we raise exception here.

File: mail/libraries/helpers.py [hunk 4]
```diff
@@ -85,6 +85,24 @@ def to_mail_message_dto(mail_data) -> EmailMessageDto:
     )
 
 
+def to_hmrc_mail_message_dto(message_id, mail_data) -> HmrcEmailMessageDto:
+    mail_contents = mail_data[1]
+    contents = b"\r\n".join(mail_contents).decode("utf-8")
```
[the remark this answers is not in the document]
Comment (saruniitr) [reply mail/libraries/helpers.py:90]: ok, used "iso-8859-1" to decode

File: mail/libraries/data_converters.py [hunk 5]
```diff
@@ -26,7 +26,10 @@ def convert_data_for_licence_data(dto: EmailMessageDto) -> dict:
         data["edi_filename"] = dto.attachment[0]
         data["edi_data"] = dto.attachment[1]
 
-    data["licence_data"]["licence_ids"] = get_licence_ids(data["edi_data"])
+    edi_data = data["edi_data"]
```
[the remark this answers is not in the document]
Comment (saruniitr) [reply mail/libraries/data_converters.py:29]: ok, used "iso8859-1" to decode

File: mock_hmrc/data_processors.py (hunk 1 above)
Comment (saruniitr) [reply mock_hmrc/data_processors.py]: used get_or_create

**Review by r4vi — APPROVED** [review]
:+1:

**Pull request merged**